src/unity/toolkits/drawing_recognition/data_preparation.cpp

```diff
@@ -0,0 +1,382 @@
+    const flex_dict &last_point = stroke[num_points-1].get<flex_dict>();
+    initialize_line(&L, first_point, last_point);
+    for (int i = 0; i < num_points; i++) {
+        float d = line_to_point_distance(&L, stroke[i].get<flex_dict>());
+        if (d > dmax) {
+            index = i;
+            dmax = d;
+        }
+    }
+    if (dmax > epsilon) {
+        flex_list first_slice = slice(stroke, 0, index-1);
+        flex_list last_slice = slice(stroke, index, num_points-1);
+        compressed_stroke = ramer_douglas_peucker(first_slice, epsilon);
+        flex_list rec_results_2 = ramer_douglas_peucker(last_slice, epsilon);
+        compressed_stroke.insert(compressed_stroke.end(), 
+            rec_results_2.begin(), rec_results_2.end());
+    } else {
+        compressed_stroke.push_back(first_point);
+        compressed_stroke.push_back(last_point);
+    }
+    return compressed_stroke;
+}
+
+flex_list simplify_drawing(flex_list raw_drawing) {
+    size_t num_strokes = raw_drawing.size();
+    size_t min_x = INT_MAX;
+    size_t max_x = 0;
+    size_t min_y = INT_MAX;
+    size_t max_y = 0;
+    flex_list simplified_drawing;
+    Point P;
+    // Compute bounds of the drawing
+    for (size_t i = 0; i < num_strokes; i++) {
+        const flex_list &stroke = raw_drawing[i].get<flex_list>();
+        for (size_t j = 0; j < stroke.size(); j++) {
+            const flex_dict &point = stroke[j].get<flex_dict>();
+            make_point(&P, point);
+            float x = P.x;
+            float y = P.y;
+            if (x < min_x) min_x = x;
+            if (x > max_x) max_x = x;
+            if (y < min_y) min_y = y;
+            if (y > max_y) max_y = y;
+        }
+    }
+    // Align the drawing to top-left corner and scale to [0,255]
+    for (size_t i = 0; i < num_strokes; i++) {
+        flex_list &stroke = raw_drawing[i].mutable_get<flex_list>();
+        flex_list new_stroke;
+        for (size_t j = 0; j < stroke.size(); j++) {
+            float new_x;
+            float new_y;
+            flex_dict &point = stroke[j].mutable_get<flex_dict>();
+            make_point(&P, point);
+            if (max_x == min_x) {
+                // vertical straight line
+                new_x = (float)min_x;
+            } else {
+                new_x = ((P.x - min_x) * 255.0) / (max_x - min_x);
+            }
+            if (max_y == min_y) {
+                // horizontal straight line
+                new_y = (float)min_y;
+            } else {
+                new_y = ((P.y - min_y) * 255.0) / (max_y - min_y);
+            }
+            flex_dict new_point;
+            new_point.push_back(std::make_pair("x", new_x));
+            new_point.push_back(std::make_pair("y", new_y));
+            new_stroke.push_back(new_point);
+        }
+        // Apply RDP Line Algorithm
+        flex_list compressed_stroke = ramer_douglas_peucker(new_stroke, 2.0);
+        simplified_drawing.push_back(compressed_stroke);
```

Comment:
May as well do one line to avoid an unnecessary copy:
```
simplified_drawing.push_back(ramen_douglas_peucker(new_stroke, 2.0f);
```
Or you can use `std::move` if you prefer two lines for clarity